[ASCII-1081] Remove Stop method from demultiplexer interface

ogaca-dd opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.

-->
### What does this PR do?
This PR removes the `Stop` method from the demultiplexer interface in favour of using fx `OnStop` hook.

This PR must be merged after https://github.com/DataDog/datadog-agent/pull/22044 to make sure `OnStop` on the demultiplexer instance is called even when there is an error.
<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation


<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes
Run the command (see below), wait 3-4 seconds (should be less than 10s) and stop the command with `ctrl+c` .  Check `Stopping aggregator` is in the log of the Agent and after you press `ctrl+c` and there is no panic.

Agent shared component
- cmd/agent/subcommands/run/command.go
- pkg/cli/subcommands/check/command.go
- cmd/dogstatsd/subcommands/start/command.go

Platform integration:
- cmd/cluster-agent/subcommands/start/command.go
- cmd/cluster-agent-cloudfoundry/subcommands/run/command.go

Open telemetry:
- cmd/otel-agent/main.go

Security Agent:
- cmd/security-agent/subcommands/start/command.go


<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

ogaca-dd committed:
Remove Run from Demultiplexer interface

ogaca-dd committed:
Add Stop in demultiplexer mock interface
defer demux.Stop(false) is not needed as done by fx.Hook

ogaca-dd committed:
Simplifying demultiplexer mock

ogaca-dd committed:
Remove Stop from demultiplexer

pgimalac reviewed (approved):
LGTM
FYI looks like you missed some windows files

ogaca-dd committed:
Do not export AgentDemultiplexer.Run

ogaca-dd commented on comp/aggregator/demultiplexer/demultiplexerimpl/demultiplexer.go:
```diff
@@ -74,6 +75,10 @@ func newDemultiplexer(deps dependencies) (provides, error) {
 	demultiplexer := demultiplexer{
 		AgentDemultiplexer: agentDemultiplexer,
 	}
+	deps.Lc.Append(fx.Hook{OnStop: func(ctx context.Context) error {
+		agentDemultiplexer.Stop(true)
+		return nil
+	}})
```
Yes, indeed, that is why I added QA instructions and carefully check it is OK to stop the aggregator later.

ogaca-dd committed:
Fix compilation on Windows

ogaca-dd committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into olivierg/remove-demux-stop

ogaca-dd committed:
Fix nil pointer issue
`Stop` is always called when using demultiplexer but NewNoopEventPlatformForwarder() set p.strategy to nil which was not check when calling `Stop`

ogaca-dd committed:
Call p.sender.Stop() only if it was started

pgimalac commented on pkg/epforwarder/epforwarder.go:
```diff
@@ -437,8 +437,10 @@ func (p *passthroughPipeline) Start() {
 func (p *passthroughPipeline) Stop() {
-	p.strategy.Stop()
-	p.sender.Stop()
+	if p.strategy != nil {
+		p.strategy.Stop()
+		p.sender.Stop()
+	}
```
Is stopping `sender` only if `strategy` is not `nil` on purpose ?

ogaca-dd replied:
`sender` calls `Start` only if `p.strategy` is not nil: https://github.com/DataDog/datadog-agent/pull/22069/files/5499155b7a536ba122a2fa189d7eabf9236abbc5..f775c5fdbea2f103f3b6e6aa438ea3406659ede7#diff-6569a3f8ce509bc4602e1db6ae7933bbf5079532cbad740507d6dfe7a6fba3f4R433-R435

pgimalac replied:
Ok perfect then, thanks for the reply !

ogaca-dd commented:
/merge

ogaca-dd commented:
/merge
